vm/src/crt_fd.rs

```diff
@@ -0,0 +1,159 @@
+    #[link_name = "_chsize_s"]
+    fn ftruncate(fd: i32, len: i64) -> i32;
+}
+#[cfg(not(windows))]
+use libc::{fsync, ftruncate};
+
+#[inline]
+fn cvt<T, I: num_traits::PrimInt>(ret: I, f: impl FnOnce(I) -> T) -> io::Result<T> {
+    if ret < I::zero() {
+        Err(crate::stdlib::os::errno())
+    } else {
+        Ok(f(ret))
+    }
+}
+
+const MAX_RW: usize = if cfg!(any(windows, target_vendor = "apple")) {
+    i32::MAX as usize
+} else {
+    isize::MAX as usize
+};
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Fd(pub i32);
+
+impl Fd {
+    pub fn open(path: &ffi::CStr, flags: i32, mode: i32) -> io::Result<Self> {
+        cvt(unsafe { libc::open(path.as_ptr(), flags, mode) }, Fd)
+    }
+
+    #[cfg(windows)]
+    pub fn wopen(path: &widestring::WideCStr, flags: i32, mode: i32) -> io::Result<Self> {
+        cvt(
+            unsafe { suppress_iph!(libc::wopen(path.as_ptr(), flags, mode)) },
+            Fd,
+        )
+    }
+
+    #[cfg(all(any(unix, target_os = "wasi"), not(target_os = "redox")))]
+    pub fn openat(&self, path: &ffi::CStr, flags: i32, mode: i32) -> io::Result<Self> {
+        cvt(
+            unsafe { libc::openat(self.0, path.as_ptr(), flags, mode) },
+            Fd,
+        )
+    }
+
+    pub fn fsync(&self) -> io::Result<()> {
+        cvt(unsafe { suppress_iph!(fsync(self.0)) }, drop)
+    }
+
+    pub fn close(&self) -> io::Result<()> {
+        cvt(unsafe { suppress_iph!(libc::close(self.0)) }, drop)
+    }
+
+    pub fn ftruncate(&self, len: i64) -> io::Result<()> {
+        cvt(unsafe { suppress_iph!(ftruncate(self.0, len)) }, drop)
+    }
+
+    /// NOTE: it's not recommended to use ManuallyDrop::into_inner() to drop the file - it won't
+    /// work on all platforms, and will swallow any errors you might want to handle.
+    #[allow(unused)] // only used on windows atm
+    pub(crate) fn as_rust_file(&self) -> io::Result<mem::ManuallyDrop<fs::File>> {
+        #[cfg(windows)]
+        let file = {
+            use std::os::windows::io::FromRawHandle;
+            let handle = self.to_raw_handle()?;
+            unsafe { fs::File::from_raw_handle(handle) }
+        };
+        #[cfg(unix)]
+        let file = {
+            let fd = self.0;
+            use std::os::unix::io::FromRawFd;
+            if fd < 0 {
+                return Err(io::Error::from_raw_os_error(libc::EBADF));
+            }
+            unsafe { fs::File::from_raw_fd(fd) }
+        };
+        #[cfg(target_os = "wasi")]
+        let file = {
+            let fd = self.0;
+            if fd < 0 {
+                return Err(io::Error::from_raw_os_error(libc::EBADF));
+            }
+            // SAFETY: as of now, File is a wrapper around WasiFd, which is a wrapper around
+            // wasi::Fd (u32). This isn't likely to change, and if it does change to a different
+            // sized integer, mem::transmute will fail.
+            unsafe { mem::transmute::<u32, fs::File>(fd as u32) }
```

Comment:
if it is same as WasiFd, is it possible to make a WasiFd value with fd here instead of transmute?

Reply:
This is just copied from the old `os::rust_file` function; WasiFd is a struct internally in the standard library and [`wasi::Fd`](https://docs.rs/wasi/0.10.2+wasi-snapshot-preview1/wasi/type.Fd.html) is a type alias.